File: 2015/23.py
# --- Day 23: Opening the Turing Lock ---

def day_23(instructions, registers):

  inst = []
  for line in instructions:
    inst.append(line.replace("\n", "").split(" "))

  counter = 0
  while counter < len(inst):
    x = inst[counter]
    if x[0] == "hlf":
      registers[x[1]] = registers[x[1]] / 2
      counter += 1
    elif x[0] == "tpl":
      registers[x[1]] = registers[x[1]] * 3
      counter += 1
    elif x[0] == "inc":
      registers[x[1]] = registers[x[1]] + 1
      counter += 1
    elif x[0] == "jmp":
      counter += int(x[1])
    elif x[0] == "jie":
      if registers[x[1].replace(",", "")] % 2 == 0:
        counter += int(x[2])
      else:
        counter += 1
    elif x[0] == "jio":
      if registers[x[1].replace(",", "")] == 1:
        counter += int(x[2])
      else:
        counter += 1

  return registers["b"]

with open("input/23.txt") as instructions:
  print(day_23(instructions, {"a": 0, "b": 0}))
with open("input/23.txt") as instructions:
  print(day_23(instructions, {"a": 1, "b": 0}))

File: 2015/23.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type instruction struct {
	inst     string
	register string
	offset   int
}

func run(instructions []instruction, a int) int {
	ptr, b := 0, 0

	for {
		if ptr >= len(instructions) {
			break
		}

		instruction := instructions[ptr]

		switch instruction.inst {
		case "hlf":
			if instruction.register == "a" {
				a = a / 2
			} else {
				b = b / 2
			}
			ptr++
			break
		case "tpl":
			if instruction.register == "a" {
				a = a * 3
			} else {
				b = b * 3
			}
			ptr++
			break
		case "inc":
			if instruction.register == "a" {
				a++
			} else {
				b++
			}
			ptr++
			break
		case "jmp":
			ptr += instruction.offset
			break
		case "jie":
			if instruction.register == "a" {
				if a%2 == 0 {
					ptr += instruction.offset
				} else {
					ptr++
				}
			} else {
				if b%2 == 0 {
					ptr += instruction.offset
				} else {
					ptr++
				}
			}
			break
		case "jio":
			if instruction.register == "a" {
				if a == 1 {
					ptr += instruction.offset
				} else {
					ptr++
				}
			} else {
				if b == 1 {
					ptr += instruction.offset
				} else {
					ptr++
				}
			}
			break
		}
	}
	return b
}

func format(line string) instruction {
	inst := instruction{"", "", 0}

	split := strings.Split(line, " ")
	inst.inst = split[0]
	if len(split) == 2 {
		if inst.inst == "jmp" {
			offset, err := strconv.Atoi(split[1])
			if err != nil {
				log.Fatal(err)
			}
			inst.offset = offset
		} else {
			inst.register = split[1]
		}
	} else {
		inst.register = string(split[1][0])
		offset, err := strconv.Atoi(split[2])
		if err != nil {
			log.Fatal(err)
		}
		inst.offset = offset
	}
	return inst
}

func main() {

	file, err := os.Open("input/23.txt")
	if err != nil {
		log.Fatal(err)
	}

	instructions := []instruction{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		instructions = append(instructions, format(scanner.Text()))
	}

	fmt.Println(run(instructions, 0))
	fmt.Println(run(instructions, 1))
}
